File: 0080_remove_duplicates_from_sorted_array_ii/solution.py
import numpy as np
import timeit
import unittest

from line_profiler import profile
from typing import List

class Solution:
    @profile
    def removeDuplicates(self, nums: List[int]) -> int:
        j = 1
        c = 1
        for i in range(1, len(nums)):
            if nums[i] == nums[j - 1]:
                # this is a repeat
                c += 1
                if c < 3:
                    # two of any one number is ok
                    if i != j:
                        nums[j] = nums[i]
                    j += 1
            else:
                # not a repeat
                if i != j:
                    nums[j] = nums[i]
                j += 1
                c = 1
        return j

    @profile
    def removeDuplicates2(self, nums: List[int]) -> int:
        j = 1
        c = 1
        n = nums[0]
        for i in range(1, len(nums)):
            if nums[i] == n:
                # this is a repeat
                c += 1
                if c < 3:
                    # two of any one number is ok
                    if i != j:
                        nums[j] = nums[i]
                    j += 1
            else:
                # not a repeat
                if i != j:
                    nums[j] = nums[i]
                j += 1
                c = 1
                n = nums[i]
        return j


class SolutionTest(unittest.TestCase):
    def test_remove_duplicates(self):
        s = Solution()
        l = [1,1,1,2,2,3]
        k = s.removeDuplicates(l)
        self.assertEqual(l[:k], [1,1,2,2,3])

        l = [1,2,3]
        k = s.removeDuplicates(l)
        self.assertEqual(l[:k], [1,2,3])

        l = [1,1,1]
        k = s.removeDuplicates(l)
        self.assertEqual(l[:k], [1,1])

        l = [0,0,1,1,1,1,2,3,3]
        k = s.removeDuplicates(l)
        self.assertEqual(l[:k], [0,0,1,1,2,3,3])


if __name__ == '__main__':
    data = np.random.randint(0, 101, 10000)
    s = Solution()
    s.removeDuplicates(data)
    # stmt = lambda: s.removeDuplicates(data)
    # time_taken = timeit.timeit(stmt, number=10000)
    # print(f"Time taken: {time_taken:.5f} seconds.")
    

File: 0080_remove_duplicates_from_sorted_array_ii/solution.cc
#include <iostream>
#include <vector>

#include <gtest/gtest.h>

using namespace std;

class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
      if (nums.size() < 3) {
        return nums.size();
      }

      int j = 2;
      for (auto i = 2; i < nums.size(); ++i) {
        if (nums[i] != nums[j - 2]) {
          nums[j++] = nums[i];
        }
      }
      return j;
    }
};

TEST(SolutionTest, shortArray) {
  Solution s;
  vector<int> v{1};
  auto k = s.removeDuplicates(v);
  ASSERT_EQ(k, 1);
  v.resize(k);
  ASSERT_EQ(v, vector<int>({1}));
}

TEST(SolutionTest, longArray) {
  Solution s;
  vector<int> v{1, 1, 1, 2, 2, 3};
  auto k = s.removeDuplicates(v);
  ASSERT_EQ(k, 5);
  v.resize(k);
  ASSERT_EQ(v, vector<int>({1, 1, 2, 2, 3}));
}
